engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailProcessor.java

```diff
@@ -0,0 +1,181 @@
+            aggregation.setBundleName(bundleName);
+            aggregation.setOrgId(orgId);
+
+            // Transform the event to
+            final JsonObject transformedEvent = this.baseTransformer.toJsonObject(event);
+            aggregation.setPayload(transformedEvent);
+
+            try {
+                this.emailAggregationRepository.addEmailAggregation(aggregation);
+            } catch (Exception e) {
+                // ConstraintViolationException may be thrown here and it must not interrupt the email that is being sent.
+                Log.warn("Email aggregation persisting failed", e);
+            }
+        }
+
+        // Fetch the template that will be used to hydrate it with the data.
+        final Optional<InstantEmailTemplate> instantEmailTemplate = this.templateRepository.findInstantEmailTemplate(event.getEventType().getId());
+        if (instantEmailTemplate.isEmpty()) {
+            // TODO don't even log
+            Log.errorf("there is no email template for this one :(");
+            return;
+        }
+
+        // Get the set of UUIDs that should receive an email notification for
+        // the given event.
+        final List<String> subscribers = this.getSubscribers(event);
+        if (subscribers.isEmpty()) {
+            // TODO don't even log
+            Log.errorf("the subscribers' UUID list is empty :/");
+            return;
+        }
+
+        final UUID historyUUID = UUID.randomUUID();
+        final Endpoint endpoint = this.endpointRepository.getOrCreateDefaultSystemSubscription(accountId, orgId, EndpointType.EMAIL_SUBSCRIPTION);
+
+        final NotificationHistory history = NotificationHistory.getHistoryStub(endpoint, event, 0L, historyUUID);
+        history.setStatus(NotificationStatus.PROCESSING);
+        this.persistNotificationHistory(history);
+
+        // Prepare all the data to be sent to the connector.
+        final EmailNotification emailNotification = new EmailNotification(
+            subscribers,
+            this.extractRecipientSettings(endpoints, event),
+            this.baseTransformer.toJsonObject(event).encode(),
+            instantEmailTemplate.get().getBodyTemplate(),
+            instantEmailTemplate.get().getSubjectTemplate(),
+            EmailSubscriptionType.INSTANT.isOptIn()
+        );
+
+        final JsonObject payload = JsonObject.mapFrom(emailNotification);
+
+        this.connectorSender.send(event, endpoint, payload);
+    }
+
+    /**
+     * Extracts the recipient settings from the given list of endpoints and
+     * the given event.
+     * @param endpoints the list of endpoints to extract the recipient settings
+     *                  from.
+     * @param event the event to extract the recipient settings from.
+     * @return a collection of recipient settings DTOs ready to be sent to the
+     * connector.
+     */
+    protected Collection<RecipientSettings> extractRecipientSettings(final List<Endpoint> endpoints, final Event event) {
```

Comment:
This method has a lot in common with this:
https://github.com/RedHatInsights/notifications-backend/blob/e506ceaa0ce8b10015d07fca028e6cad16ccbda9/engine/src/main/java/com/redhat/cloud/notifications/processors/SystemEndpointTypeProcessor.java#L34-L37

Is it possible to reuse the existing code to avoid duplicating the code?

Reply:
Extracted to a common function!